### PR TITLE
[FIRRTL] canonicalize zero-width memories

### PR DESCRIPTION
**Pull request by darthscsi**

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -1185,3 +1185,25 @@ void NodeOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    for (auto user : port.getUsers())
+      if (!isa<SubfieldOp>(user))
+        return failure();
+
+  // Annoyingly, there isn't a good replacement for the port as a whole, since
+  // they have an outer flip type.
+  for (auto port : op->getResults()) {
+    for (auto user : llvm::make_early_inc_range(port.getUsers())) {
+      SubfieldOp sfop = cast<SubfieldOp>(user);
+      rewriter.replaceOpWithNewOp<WireOp>(sfop, sfop.result().getType());
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1204]: Ok, I guess this makes sense and composes with connect elimination.

**Review by lattner — APPROVED** [review]
nice

File: test/Dialect/FIRRTL/canonicalization.mlir [hunk 2]
```diff
@@ -1440,5 +1440,53 @@ firrtl.module @MuxInvalidTypeOpt(in %in : !firrtl.uint<1>, out %out : !firrtl.ui
     %c9_si10 = firrtl.constant 9 : !firrtl.sint<10>
     firrtl.partialconnect %y, %c9_si10 : !firrtl.sint<4>, !firrtl.sint<10>
   }
+
+firrtl.module @zeroWidthMem(in %clock: !firrtl.clock) {
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply test/Dialect/FIRRTL/canonicalization.mlir:1444]: It's there.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1204]: One can eliminate all the connects too, but you increase the number of weird things that may be happening and prevent triggering.

File: test/Dialect/FIRRTL/canonicalization.mlir (hunk 2 above)
[the remark this answers is not in the document]
Comment (darthscsi) [reply test/Dialect/FIRRTL/canonicalization.mlir:1444]: It's there at the bottom, but I'll move it.

**Review by lattner — APPROVED** [review]
oh thanks, I missed it, looks great

**Pull request merged**